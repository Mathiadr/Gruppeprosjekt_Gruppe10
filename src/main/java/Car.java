import java.util.ArrayList;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether or not the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    private String fuelType;        //TODO: IMPLEMENT
    private String transmission;    //TODO: IMPLEMENT

    private ArrayList<String> features;


    private static ArrayList<Car> carArrayList = new ArrayList<>();

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }


    public String getAvailableDate() {
        return availableDate;
    }

    public void setAvailableDate(String availableDate) {
        this.availableDate = availableDate;
    }

    public static ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public static void setCarArrayList(ArrayList<Car> carArrayList) {
        Car.carArrayList = carArrayList;
    }

    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
