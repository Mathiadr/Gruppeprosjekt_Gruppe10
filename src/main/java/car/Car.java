package car;

import java.util.ArrayList;
import java.util.Collection;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    //private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    //private String fuelType;        //TODO: IMPLEMENT
    //private String transmission;    //TODO: IMPLEMENT

    //private Collection<String> features; //TODO: IMPLEMENT

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
