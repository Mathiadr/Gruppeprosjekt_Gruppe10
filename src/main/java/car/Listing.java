package car;

import java.util.Date;
import java.time.format.DateTimeFormatter;

public class Listing {
        // Contains information of every bid, each of which belongs to one car.
        // TODO: EVERY CAR BELONGS TO ONE LISTING
        // TODO: ur mum
        private String publisherUsername;
        private Date startDate;
        private Date endDate;
        private boolean isAvailable;
        private boolean published;
        private String description;


        public Listing(){}

        public Listing(String publisherUsername, Date startDate, Date endDate, boolean published, String description){
            this.publisherUsername = publisherUsername;
            this.startDate = startDate;
            this.endDate = endDate;
            this.isAvailable = true;
            this.published = published;
            this.description = description;


        }


        public Date getStartDate() {
            return startDate;
        }

        public void setStartDate(Date startDate) {
            this.startDate = startDate;
        }

        public Date getEndDate() {
            return endDate;
        }

        public void setEndDate(Date endDate) {
            this.endDate = endDate;
        }

        public boolean isAvailable() {
            return isAvailable;
        }

        public void setAvailable(boolean available) {
            isAvailable = available;
        }

        public boolean isPublished() {
            return published;
        }

        public void setPublished(boolean published) {
            this.published = published;
        }
}

