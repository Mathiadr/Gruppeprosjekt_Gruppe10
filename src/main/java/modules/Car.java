package modules;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Date;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String brand;                   //TODO: IMPLEMENT
    //private String fuelType;              //TODO: IMPLEMENT
    //private String transmission;          //TODO: IMPLEMENT
    //private Collection<String> features;  //TODO: IMPLEMENT
    private Listing listing;

    public static final String[] fuelTypesList = {"Electric", "Gasoline", "Diesel"};
    public static final String[] transmissionList = {"Manual", "Automatic"};
    public static final String[] featureList = {"GPS", "Air Conditioning"};



    public Car(){

    }

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        LocalDate startDateplaceholder = LocalDate.of(2022, 1, 1);
        LocalDate endDatePlaceholder = LocalDate.of(2022, 12, 31);
        this.listing = new Listing(startDateplaceholder, endDatePlaceholder, true, "");
    }
    public Car(String registrationNumber, String owner, String model, Listing listing) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        this.listing = listing;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public Listing getListing() {
        return listing;
    }

    public void setListing(Listing listing) {
        this.listing = listing;
    }



    @Override
    public String toString(){
        return "\n\nregistrationNumber: " + registrationNumber + ",\n" +
                "Owner: " + owner + ",\n" +
                "Model: " + model + ",\n" +
                "Listing: " + listing;
    }
}
