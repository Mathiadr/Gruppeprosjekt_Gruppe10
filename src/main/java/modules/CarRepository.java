package modules;

import tools.CarRepositoryFileHandler;

import java.io.File;
import java.util.ArrayList;

public class CarRepository {
    private String repositoryName; // This holds the name of the repository being accessed.
    private ArrayList<Car> carArrayList = new ArrayList<>();
    private final CarRepositoryFileHandler carRepositoryFileHandler = new CarRepositoryFileHandler();

    public CarRepository(String repositoryName) {
        this.repositoryName = repositoryName;

        carArrayList = readFromJSON();
    }



    // Adds a new car to the repository file
    public void AddNewCar(Car car){
        //TODO: implement
        boolean exists = false;
        for (Car i : carArrayList){
            //Checks if the car being added does not already exist within the list
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                exists = true;
                break;
            }
        }
        if (!exists){
            //System.out.println("Adding car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " to carArrayList");
            this.carArrayList.add(car);
        }
        else{
            //System.out.println("Car " + car.getRegistrationNumber() + " Already exists");
        }
    }

    // Removes Car object from carArrayList
    public void RemoveExistingCar(Car car){
        if (carArrayList.contains(car)){
            //System.out.println("Removed car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " from carArrayList");
            carArrayList.remove(car);
        }
        else{
            // TODO: Implement error message to GUI
            //System.out.println("Could not find car " + car.getRegistrationNumber() + " in CarArrayList");
        }
    }

    // It rents da car, dumbass
    public void RentCar(Car car){
        // TODO: Test if works
        try{
            for (Car i : GetAllAvailableCars()){
                if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                    //TODO: ADD EXTRA MEASURE TO ENSURE IT DOES NOT RENT AN UNAVAILABLE CAR
                    i.getListing().setAvailable(false);
                }
            }
        }
        catch (RuntimeException couldNotRentCar){
            //System.out.println("\n\nCould not rent car " + car.getRegistrationNumber() + "\n\n");
        }

    }

    // Returns all cars that are not currently occupied
    public ArrayList<Car> GetAllAvailableCars(){
        ArrayList<Car> allAvailableCars = new ArrayList<>();
        for (Car i : carArrayList){
            if (i.getListing().isAvailable()){
                allAvailableCars.add(i);
            }
        }
        return allAvailableCars;
    }

    public void updateListing(Car car, Listing listing){
        for (Car i : carArrayList){
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                i.setListing(listing);
                //System.out.println("Updated listing in Car " + i.getRegistrationNumber());
                break;
            }
        }
    }

    public void SaveCarsToJSON(){
        carRepositoryFileHandler.writeArrayListToFile(getCarArrayList(), repositoryName);
    }

    // Reads all objects from JSON file and maps it to the Car object
    public ArrayList<Car> readFromJSON(){
        return carRepositoryFileHandler.readFromFile(repositoryName);
    }

    public String getRepositoryName() {
        return repositoryName;
    }

    public void setRepositoryName(String repositoryName) {
        this.repositoryName = repositoryName;
    }

    public ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public void setCarArrayList(ArrayList<Car> carArrayList) {
        this.carArrayList = carArrayList;
    }
}
