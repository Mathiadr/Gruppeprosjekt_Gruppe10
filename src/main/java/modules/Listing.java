package modules;


import java.util.Date;

public class Listing {
    // Contains information of every bid, each of which belongs to one car.
    // TODO: EVERY CAR BELONGS TO ONE LISTING
    // TODO: ur mum
    private String publisherUsername;
    private Date startDate;
    private Date endDate;
    private Car car;
    private boolean isAvailable;
    private boolean published;
    private String description;

    public Listing(){}

    public Listing(String publisherUsername, Date startDate, Date endDate, Car car, boolean published, String description){
        this.publisherUsername = publisherUsername;
        this.startDate = startDate;
        this.endDate = endDate;
        this.car    = car;
        this.isAvailable = true;
        this.published = published;
        this.description = description;
    }


    public Date getStartDate() {
        return startDate;
    }

    public void setStartDate(Date startDate) {
        this.startDate = startDate;
    }

    public Date getEndDate() {
        return endDate;
    }

    public void setEndDate(Date endDate) {
        this.endDate = endDate;
    }

    public Car getCar() {
        return car;
    }

    public void setCar(Car car) {
        this.car = car;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }

    public boolean isPublished() {
        return published;
    }

    public void setPublished(boolean published) {
        this.published = published;
    }



}
