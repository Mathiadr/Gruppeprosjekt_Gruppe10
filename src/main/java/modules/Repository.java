package modules;

import tools.RepositoryFileHandler;

import java.util.ArrayList;
import java.util.Objects;

public class Repository {
    private String repositoryName; // This holds the name of the repository being accessed.
    private ArrayList<Listing> listingArrayList = new ArrayList<Listing>();

    private final RepositoryFileHandler repositoryFileHandler = new RepositoryFileHandler();

    public Repository(String repositoryName) {
        this.repositoryName = repositoryName;

        listingArrayList = readFromJSON();
    }



    // Reads all objects from JSON file and returns an ArrayList of Listing objects
    public ArrayList<Listing> readFromJSON(){
        return repositoryFileHandler.readFromFile(repositoryName);
    }


    public void addNewListing(Listing listing){
        boolean exists = false;
        for (Listing i : listingArrayList){
            //TODO: IMPLEMENT
            if(Objects.equals(i.getCar().getRegistrationNumber(), listing.getCar().getRegistrationNumber())){
                exists = true;
                break;
            }
        }
        if(!exists){
            listingArrayList.add(listing);
        }

    }

    // Removes Car object from listingArrayList
    public void removeExistingListing(Car car){
        // TODO: IMPLEMENT
    }

    // It rents da car, dumbass
    public void acceptListing(Car car){
        //TODO: REMOVE & RENAME
    }

    // Returns all cars that are not currently occupied
    public ArrayList<Listing> GetAllAvailableListings(){
        // TODO: Add date as parameter argument
        ArrayList<Listing> allAvailableListings = new ArrayList<>();
        for (Listing i : listingArrayList){
            if (i.isAvailable()){
                allAvailableListings.add(i);
            }
        }
        return allAvailableListings;
    }

    public void SaveCarsToJSON(){
        repositoryFileHandler.writeArrayListToFile(getListingArrayList(), repositoryName);
    }

    public String getRepositoryName() {
        return repositoryName;
    }

    public void setRepositoryName(String repositoryName) {
        this.repositoryName = repositoryName;
    }

    public ArrayList<Listing> getListingArrayList() {
        return listingArrayList;
    }

    public void setListingArrayList(ArrayList<Listing> listingArrayList) {
        this.listingArrayList = listingArrayList;
    }
}
