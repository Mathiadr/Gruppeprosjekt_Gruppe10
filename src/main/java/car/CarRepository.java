package car;

import org.json.JSONObject;

import java.io.*;
import java.util.ArrayList;

public class CarRepository {
    private String repositoryName;

    private static ArrayList<Car> availableCarArrayList; // Holds cars not currently occupied
    private static ArrayList<Car> allCarArrayList; // Holds ALL cars. Unsure if necessary

    CarRepository(String repositoryName) throws FileNotFoundException {
        this.repositoryName = repositoryName;

        FileReader fileReader = new FileReader(this.repositoryName);
    }

    // Adds a new car to the repository file
    public void AddNewCar(Car car){
        //TODO: implement
        JSONObject jsonObject = new JSONObject(car.getRegistrationNumber());
        jsonObject.put("registrationNumber", car.getRegistrationNumber());
        jsonObject.put("owner", car.getOwner());
        jsonObject.put("model", car.getModel());
        jsonObject.put("availableDate", car.getAvailableDate());
        jsonObject.put("isAvailable", car.isAvailable());
        jsonObject.put("fuelType", car.getFuelType());
        jsonObject.put("transmission", car.getTransmission());
        jsonObject.put("features", car.getFeatures()); //Maybe a nested JSONObject/JSONArray belongs here?
        try{
            FileWriter fileWriter = new FileWriter(this.repositoryName);
            fileWriter.append(jsonObject.toString());
            System.out.println("Car " + car.getRegistrationNumber() + " has been written to file.");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // It rents da car, dumbass
    public void RentCar(Car car){
        //TODO: implement
    }

    // Returns all cars that are not currently occupied
    public ArrayList<Car> GetAllAvailableCars(){
        //TODO: implement
        return null;
    }

    // Unsure what to do here...
    public void SaveAllAvailableCars(){
        //TODO: implement
    }


}
