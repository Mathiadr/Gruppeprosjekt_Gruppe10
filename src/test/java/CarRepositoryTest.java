import modules.Car;
import modules.CarRepository;
import modules.Listing;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;

public class CarRepositoryTest {
    @Test
    public void new_car_is_created_and_added_to_ArrayList(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo");
        carRepository.AddNewCar(car1);
        assertTrue(carRepository.getCarArrayList().contains(car1));
    }

    @Test
    public void JSON_writes_to_ArrayList_and_is_of_class_Car(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        carRepository.readFromJSON();
        ArrayList<Car> carArrayList = carRepository.getCarArrayList();
        assertFalse(carArrayList.isEmpty());
        assertSame(carArrayList.get(0).getClass(), Car.class);
    }

    @Test
    public void JSON_does_not_contain_multiple_similar_registrationNumbers(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo");
        Car car2 = new Car("DFG441563", "Dummy two", "Toyota");
        carRepository.AddNewCar(car1);
        carRepository.AddNewCar(car2);
        ArrayList<Car> carArrayList = carRepository.getCarArrayList();
        assertTrue(carArrayList.contains(car1), "Specified car was not added");
        assertFalse(carArrayList.contains(car2), "Two cars with the same registrationNumber exist");
    }

    @Test
    public void Car_is_removed_from_carArrayList(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo");
        ArrayList<Car> carArrayList = carRepository.getCarArrayList();
        carRepository.AddNewCar(car1);
        assertTrue(carArrayList.contains(car1), "Specified car was not added");
        carRepository.RemoveExistingCar(car1);
        assertFalse(carArrayList.contains(car1), "Specified car was not removed");
    }

    @Test
    public void available_cars_list_does_not_contain_unavailable_cars(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo");
        Car car2 = new Car("AFGH13543", "Dummy two", "Toyota");
        Listing listing1 = new Listing(null, null, true, null);
        Listing listing2 = new Listing(null, null, false, null);
        listing1.setAvailable(true);
        listing2.setAvailable(false);
        car1.setListing(listing1);
        car2.setListing(listing2);
        carRepository.AddNewCar(car1);
        carRepository.AddNewCar(car2);
        ArrayList<Car> availableCarsArrayList = carRepository.GetAllAvailableCars();
        assertTrue(availableCarsArrayList.contains(car1), "Available car returned as not available");
        assertFalse(availableCarsArrayList.contains(car2), "Unavailable car returned as available");
    }

    @Test
    public void car_listing_is_updated(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo",
                new Listing(null, null, true, "This is the original listing"));
        carRepository.AddNewCar(car1);
        Listing updatedListing = new Listing(null, null, true, "This is the updated listing");
        carRepository.updateListing(car1, updatedListing);
        assertNotEquals("This is the original listing", car1.getListing().getDescription());
        assertEquals("This is the updated listing", car1.getListing().getDescription());
    }


    @Test
    public void car_becomes_unavailable_when_renting(){
        CarRepository carRepository = new CarRepository("testRepository.json");
        Car car1 = new Car("DFG441563", "Dummy one", "Volvo",
                new Listing(null, null, true, "TEST DESCRIPTION"));
        carRepository.AddNewCar(car1);
        carRepository.RentCar(car1);
        assertFalse(car1.getListing().isAvailable());
    }
}
