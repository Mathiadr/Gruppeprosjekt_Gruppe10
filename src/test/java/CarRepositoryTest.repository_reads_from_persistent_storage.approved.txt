[

registrationNumber: MXH10045,
Owner: Mathias Dale Ratdal,
Model: E2,
Listing:    Start date: null,
   End date: null,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: EBX23365,
Owner: Sokrat Starynkiewicz,
Model: A4,
Listing:    Start date: null,
   End date: null,
   isAvailable: false,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: E0B10045,
Owner: Joe Mama,
Model: C2,
Listing:    Start date: null,
   End date: null,
   isAvailable: false,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: ADHD0O11,
Owner: Sander Kristiansen,
Model: C6,
Listing:    Start date: null,
   End date: null,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: D1CK9864,
Owner: Djikstra Van Houven,
Model: BA1,
Listing:    Start date: null,
   End date: null,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: P0043456,
Owner: John Doe,
Model: A2,
Listing:    Start date: null,
   End date: null,
   isAvailable: false,
   Description : PLACEHOLDER DESCRIPTION, 

registrationNumber: BE240051,
Owner: Ola Nordmann,
Model: A1,
Listing:    Start date: null,
   End date: null,
   isAvailable: true,
   Description : PLACEHOLDER DESCRIPTION]